### DISCUSSION TITLE
`Disasm_expert.Basic.available_backends ()` shows no backends

### DISCUSSION THREAD

**Discussion by Cryspia** [BinaryAnalysisPlatform/bap#1583] [Q&A]
I am trying to build a BIL lifter server using the Bap.Std API (v2.5.0) that reads in machine code with their memeory addresses and output BIL strings, but failed at a very first stage.
I have successfully created a test mem variable consists of "90c3" (nop; ret) and an "x86_64" arch.
I tried to disassemble the bytes using Disasm.of_mem and Disasm.insns, but got an empty list.
Then I tried to use the low level API by calling a Disasm_expert.Linear.sweep, but got an exception "Unknown backend: llvm"
I tried to check all availble backends using Disasm_expert.Basic.available_backends (), but got en empty list.

I am confused about the outcomes, and cannot figure out in which step did I make a mistake. I have scanned the API manual, but did not found useful hints.

Can anyone enlighten me or point me to a locaiton of the document?

BTW: I cannot find a way to transfer disassembled instructions into BIL by looking at the BIL API functions. Could anyone teach me which API should I use?

**Comment by Cryspia** [comment]
Tried on both a MacOS and a ubuntu20.04 docker container. Checked that `bap-llvm` is installed. Still have not get the luck.

**Comment by ivg** [comment]
I can guess that you were writing a standalone program (not a bap plugin) and using BAP as a library. In that case you have to call `Bap_main.init` before doing any calls to BAP. See [documentation][1] for more info.

> BTW: I cannot find a way to transfer disassembled instructions into BIL by looking at the BIL API functions. Could anyone teach me which API should I use?

`Disasm.of_mem` will return a disassembled and lifted (equipped with semantics) instruction represented with `Insn.t`. You can use `Insn.bil` to access the BIL representation of semantics. 

[1]: http://binaryanalysisplatform.github.io/bap/api/master/bap-main/Bap_main/index.html
